[libSyntax] Incremental Syntax Parsing

ahoppen opened the pull request:
This is a work in progress PR to add incremental syntax parsing to the Swift compiler. It accompanies the proposal [here](https://gist.github.com/ahoppen/3ae1a6cd64e558710a4afcd372e8fdc4) and discussion thread [here](https://forums.swift.org/t/incremental-syntax-parsing/12368).

ahoppen committed:
[libSyntax] Documentation improvements

ahoppen committed:
[libSyntax] Add static_asserts for the size of RawSyntaxBits

ahoppen committed:
[libSyntax] Update space needed for raw syntax bits

The bit for ManualMemory was not taken care of previously

ahoppen committed:
[libSyntax] Compute the text length of every node on the fly

ahoppen committed:
Wrap the path to -sdk in quotes for %target-swift-frontend

ahoppen committed:
[incrParse] Reparse a node if the next leaf node has been modified

ahoppen committed:
[swift-syntax-test] Fix formatting of command-line-argument descriptions

ahoppen committed:
[incrParse] Allow information about node reused be outputted

ahoppen committed:
[incrParse] Refactor node reusability into separate function

ahoppen committed:
[incrParse] Allow line:column notation to specify edits

ahoppen committed:
[incrParse] Store reused regions and output them after parsing

ahoppen committed:
[incrParse] Add coloured output indicating which code got reused

ahoppen committed:
[incrParse] Support incremental parsing for edited files

ahoppen committed:
[swift-syntax-test] Refactor to allow incremental parsing on all actions

ahoppen committed:
[incParse] Make the SytnaxParsingCache operate on the leading trivia's start

ahoppen committed:
[incrParse] Add utility to test incremental parsing

ahoppen committed:
[incrParse] Compute byte offsets of pre-edit file based on that file

ahoppen committed:
[incrParse] Reparse nodes if the next node's trailing trivia has changed

ahoppen committed:
[incrParse] Add some more passing tests

harlanhaskins commented on lib/Parse/SyntaxParsingCache.cpp:
```diff
@@ -15,6 +15,45 @@
 using namespace swift;
 using namespace swift::syntax;
 
+ArrayRef<TriviaPiece> getNextChildsLeadingTrivia(const Syntax *Node) {
```
Would be nice to make this `getNextChild`, because that's a traversal I've found very necessary.

ahoppen replied:
Sounds good. I will keep it in mind when I refactor these methods to provide a (hopefully faster) cache based lookup.

ahoppen committed:
[incrParse] Outdated documentation fixes

ahoppen committed:
[incrParse] Add option to force coloured output

ahoppen committed:
[incrParse] Add option to the test utility to print visual reuse info

ahoppen committed:
[incrParse] Test utility: Put swift-syntax-test args in quotes if needed

ahoppen committed:
[incrParse] Add verification of reparsed regions to swift-syntax-test

ahoppen committed:
[incrParse] Allow testing of multiple edits per line

ahoppen committed:
[incrParse] Add support for test cases that verify the reused regions

ahoppen committed:
[incrParse] Allow whitespaces to be reparsed in test

This greatly improves the ergonomics of writing tests and outweighs the
ability to test which whitespaces get parsed since their parsing
overhead should be minimal.

ahoppen committed:
[incrParse] Allow reuse of MemberDeclListItems

ahoppen committed:
[incrParse] Fix parsing of nodes covering no source text

ahoppen committed:
[incrParse] Add some more tests

ahoppen committed:
[libSyntax] Add a debug dump function to SyntaxParsingContext

ahoppen committed:
[incrParse] Fix lexer offset issue when missing tokens get synthesized

ahoppen committed:
[libSyntax] Add support for parsing #sourceLocation directives

ahoppen committed:
[incrParse] Minor improvements to the test utility

ahoppen committed:
[libSyntax] Omit unknown nodes if they do not have any children

ahoppen committed:
[incrParse] Fix swift-syntax-test not complaining about unexpected reparsed functions

ahoppen committed:
[libSyntax] Fix parsing of StringLiterals with invalid interpolation segments

ahoppen committed:
[incrParse] Reparse a node if the next leaf node has been modified

ahoppen committed:
[incrParse] Add error nodes to the end of a CodeBlockItem

Creating a new CodeBlockItem meant that when doing an edit in the error
nodes, the prefix gets reused and thus the code is parsed as invalid
although it is not.

ahoppen committed:
[libSyntax] Improve syntax related dump functions

ahoppen committed:
[incrParse] Aesthetic improvements to the test utility

ahoppen committed:
[libSyntax] Fix SwiftSyntax CodeBlockItemList test

ahoppen committed:
[libSyntax] Adjust tests for improved syntax parsing behaviour

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test

nkcsgexi commented on include/swift/Syntax/RawSyntax.h:
```diff
@@ -434,6 +442,28 @@ class RawSyntax final
   }
 
+  /// Return the number of bytes this node takes when spelled out in the source
+  size_t getTextLength() {
```
can we re-use `accumulateAbsolutePosition` to calculate the length?

ahoppen replied:
I'm using `accumulateAbsolutePosition` now to compute the length of a token. For all other nodes I'm still computing the value while constructing the tree because it is faster.

nkcsgexi commented on include/swift/Parse/SyntaxParsingCache.h:
```diff
@@ -0,0 +1,101 @@
+  llvm::Optional<Syntax> lookUp(size_t NewPosition, SyntaxKind Kind);
+
+  /// Turn recording of reused ranges on
+  void recordReuseInformation() { RecordReuseInformation = true; }
```
A better name is `setRecordReuseInformation()`.

nkcsgexi commented on include/swift/Parse/SyntaxParsingContext.h:
```diff
@@ -112,6 +113,9 @@ class alignas(1 << SyntaxAlignInBits) SyntaxParsingContext {
   SyntaxParsingContext *&CtxtHolder;
 
+  /// A cache of nodes that can be reused when creating the current syntax tree
+  SyntaxParsingCache *SyntaxParsingCache = nullptr;
```
We could move this instance to `RootContextData`. It will make it clear that all `SyntaxParsingContext` share a single cache entry. Same with `Arena`.

nkcsgexi commented on lib/Parse/SyntaxParsingContext.cpp:
```diff
@@ -37,13 +37,34 @@ SyntaxParsingContext::SyntaxParsingContext(SyntaxParsingContext *&CtxtHolder,
+    return 0;
+  }
+  // Disable this node and push the cached nodes into the storage
+  disable();
```
Instead of calling `disable()`, can we define another `AccumulationMode` called `LoadedFromCache` to be more specific? Probably, we can assert `LoadedFromCache` context doesn't spawn any child context.

nkcsgexi reviewed (commented):
Looking good overall! Some comments so far.

ahoppen commented on include/swift/Syntax/SyntaxKind.h.gyb:
```diff
@@ -71,6 +71,8 @@ bool isTokenKind(SyntaxKind Kind);
 
 bool isUnknownKind(SyntaxKind Kind);
 
+bool shallBeOmittedWhenNoChildren(SyntaxKind Kind);
```
I didn't want to create a gyb file in the parser for this. It is currently still nice and gyb-free.

Maybe we could add this function to the `swift` namespace instead of the `swift::syntax` one and call it `parserShallOmitIfNoChildern`

nkcsgexi replied:
yeah, it works too, as long as we make it explicit that it's a parser thing rather than for syntax nodes in general.

ahoppen committed:
[libSyntax] Rename recordReuseInformation to setRecordReuseInformation

ahoppen committed:
[libSyntax] Store shared SyntaxParsingContext data in RootContextData

ahoppen commented:
@swift-ci Please smoke test

rintaro commented on lib/Parse/Parser.cpp:
```diff
@@ -709,6 +709,23 @@ void Parser::skipUntilConditionalBlockClose() {
+    L->advanceTo(Tok, LeadingTrivia, TextLength);
+    L->lex(Tok, LeadingTrivia, TrailingTrivia);
+    return true;
+  }
```
I feel this needlessly complicated.
How about:
```cpp
  unsigned Offset = SourceMgr.getLocOffsetInBuffer(Tok.getLoc(), L->getBufferID());
  Offset -= LeadingTrivia.getTextLength();
  if (auto TextLength = SyntaxContext->loadFromCache(Offset)) {
    L->resetToOffset(Offset + TextLength);
    L->lex(Tok, LeadingTrivia, TrailingTrivia);
    return true;
  }
```
where `Lexer:: resetToOffset` is just:
```swift
Lexer::resetToOffset(size_t Offset) {
  assert(BufferStart + Offset <= BufferEnd);
  CurPtr = BufferStart + Offset;

  lexImpl();
}
```

rintaro commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -3349,6 +3352,8 @@ ParserResult<PoundDiagnosticDecl> Parser::parseDeclPoundDiagnostic() {
 
 ParserStatus Parser::parseLineDirective(bool isLine) {
+  SyntaxParsingContext PoundSourceLocation(SyntaxContext,
+                                           SyntaxKind::PoundSourceLocation);
```
Good catch!
Could you separate `#sourceLocation` parsing out of this PR if possible?

rintaro commented on test/lit.cfg:
```diff
@@ -721,7 +722,7 @@ if run_vendor == 'apple':
     config.target_ld = "%s ld -L%r" % (xcrun_prefix, make_path(test_resource_dir, config.target_sdk_name))
     config.target_swift_frontend = (
-        "%s -frontend %s -sdk %s %s %s" %
+        "%s -frontend %s -sdk '%s' %s %s" %
```
Unrelated change?

rintaro reviewed (commented):
Looks pretty good overall.
Some comments, nothing critical.

ahoppen commented on lib/Parse/SyntaxParsingCache.cpp:
```diff
@@ -0,0 +1,94 @@
+    auto NextRawNode = NextLeafNode->getRaw();
+    NextLeafNodeLength += NextRawNode->getTextLength();
+    for (auto TriviaPiece : NextRawNode->getLeadingTrivia()) {
+      NextLeafNodeLength += TriviaPiece.getTextLength();
```
This is a bug. I fixed it in the syntax colouring branch but missed to cherry-pick it over. Done that now.

ahoppen commented on include/swift/Parse/SyntaxParsingContext.h:
```diff
@@ -174,20 +183,41 @@ class alignas(1 << SyntaxAlignInBits) SyntaxParsingContext {
   ~SyntaxParsingContext();
 
+  /// Try loading the current node from the \c SyntaxParsingCache by looking up
+  /// if an unmodified node exists at \c LexerOffset of the same kind. If a node
```
Thanks. I didn't know that.

ahoppen commented:
@swift-ci Please smoke test

harlanhaskins commented:
🎉🎉🎉
